### PR TITLE
musly: add patches for FFmpeg 7, C++17, and external deps; kissfft: build with CMake; libresample: 0.1.3 -> 0.1.4-unstable-2024-08-23

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

More time spent on software that probably nobody cares about at this point. One fewer FFmpeg 4 dependency in the world, and a couple more packages that work on macOS!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by ggPeti — APPROVED** [review]

**Comment by emilazy** [comment]
Against all odds, libresample upstream… is accepting my pull request to add a complete rewrite of the build system!

Drafting this for now until I can get it merged and integrated.

**Comment by emilazy** [comment]
Result of `nixpkgs-review pr 332035` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>ardour</li>
    <li>kissfft</li>
    <li>kissfft.bin</li>
    <li>kissfft.dev</li>
    <li>kissfftFloat</li>
    <li>kissfftFloat.bin</li>
    <li>kissfftFloat.dev</li>
    <li>libresample</li>
    <li>libresample.bin</li>
    <li>libresample.dev</li>
    <li>loudgain</li>
    <li>musly</li>
    <li>musly.bin</li>
    <li>musly.dev</li>
    <li>musly.doc</li>
    <li>qm-dsp</li>
    <li>sonic-pi</li>
    <li>zrythm</li>
  </ul>
</details>

Result of `nixpkgs-review pr 332035` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>kissfft</li>
    <li>kissfft.bin</li>
    <li>kissfft.dev</li>
    <li>kissfftFloat</li>
    <li>kissfftFloat.bin</li>
    <li>kissfftFloat.dev</li>
    <li>libresample</li>
    <li>libresample.bin</li>
    <li>libresample.dev</li>
    <li>loudgain</li>
    <li>musly</li>
    <li>musly.bin</li>
    <li>musly.dev</li>
    <li>musly.doc</li>
    <li>qm-dsp</li>
  </ul>
</details>

File: pkgs/by-name/li/libresample/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/li/libresample/package.nix:63]: We add or licenses as both

File: pkgs/by-name/li/libresample/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/li/libresample/package.nix:70]: Wrong commit

File: pkgs/by-name/ki/kissfft/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/ki/kissfft/package.nix:21]: Then just drop it instead. No need to maintain compatibility code for overrides that are likely unused

**Comment by ggPeti** [comment]
Tried musly with a bunch of audio files in the following way:
```
musly -N # initialize collection
musly -a . # add audio files from the current directory
musly -m sim # print similarity matrix to a file
```
and it seems to work fine.

**Review by ggPeti — APPROVED** [review]

File: pkgs/by-name/li/libresample/package.nix
Comment (emilazy) [reply pkgs/by-name/li/libresample/package.nix:64]: Please see my [previous response](https://github.com/NixOS/nixpkgs/pull/330741#pullrequestreview-2203607963) (and [the one I linked last time](https://github.com/NixOS/nixpkgs/pull/329470#discussion_r1688547936) with more thoughts from others). It adds value, does no harm, and the argument that it is redundant also applies to other possible settings of `sourceProvenance`. Nobody objects to the equivalent `license = lib.licenses.free;`. (Anyway, just like licence information and main programs, we should be moving in the direction of requiring explicit `sourceProvenance` for all packages, not away from it.)

File: pkgs/by-name/li/libresample/package.nix
Comment (emilazy) [reply pkgs/by-name/li/libresample/package.nix:63]: As far as I can tell, this would make any software bill of materials for a Nix project using `libresample` strictly less useful, given that LGPL 2.1+’s requirements are a superset of BSD3’s, and there are examples of doing it the other way in‐tree (I suspect many dual licenses just aren’t represented at all, actually). However if you want me to make it less useful I can.

File: pkgs/by-name/li/libresample/package.nix
Comment (emilazy) [reply pkgs/by-name/li/libresample/package.nix:70]: No, it’s the correct commit. The tools were not being built before.

File: pkgs/by-name/ki/kissfft/package.nix
Comment (emilazy) [reply pkgs/by-name/ki/kissfft/package.nix]: Done.

**Comment by smancill** [comment]
Result of `nixpkgs-review pr 332035` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>zrythm</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>kissfft</li>
    <li>kissfft.bin</li>
    <li>kissfft.dev</li>
    <li>kissfftFloat</li>
    <li>kissfftFloat.bin</li>
    <li>kissfftFloat.dev</li>
    <li>libresample</li>
    <li>libresample.bin</li>
    <li>libresample.dev</li>
    <li>musly</li>
    <li>musly.bin</li>
    <li>musly.dev</li>
    <li>musly.doc</li>
    <li>qm-dsp</li>
  </ul>
</details>

**Comment by smancill** [comment]
Looks ok to me, although that's a lot of patches to maintain.

My only gripe is that I think we shouldn't be merging branches with random names.

**Comment by emilazy** [comment]
The `libresample` Meson patch has been [provisionally accepted upstream](https://github.com/minorninth/libresample/pull/8#issuecomment-2267148311) and I’m just waiting on final review. As soon as a new release is tagged I’ll update the package and drop the patches we have here. So there shouldn’t be any ongoing maintenance cost there and the new build system has already been incorporated into the Debian package.

The `kissfft` one I also sent a PR for, but no idea if or when it will be merged. It’s a one line patch that is smaller than the patch we were having to carry for the Make build system, so I think it’s fine.

The `musly` ones… yeah, no argument there. Upstream hasn’t done anything in half a decade, so it’s very unlikely the patches will require any ongoing maintenance cost. I have no idea if they’ll merge my PR or not. Given that it was already broken with a newer C++ standard and FFmpeg 4 is very close to completely removed from Nixpkgs now the alternative would probably be to drop the leaf package. Since the patches are already written and @ggPeti has been responsive as a maintainer I don’t think there’s any need to consider doing that at the present juncture, but I wouldn’t stand in the way of its removal either.

The branch name is a [Jujutsu](https://github.com/martinvonz/jj) change ID and makes pushing out changes a lot easier for me. I know I’m not the only Nixpkgs contributor using Jujutsu, although I may be the most prolific. Agreed that they don’t make for fantastic merge commit subject lines, but scanning the merge log of Nixpkgs is kind of hopeless anyway, and I’ve seen many merged PRs with manually‐written branch names that are somewhere between unhelpful and actively inaccurate. The PR title is included in the second line of the merge commit messages, which is a much more sensible way to browse the merge history, but I do wish that GitHub would include it in the subject line instead. If you’re implacably opposed I can attach another branch name, but it would require closing this PR and opening another so I’m not sure it’s exactly worth it.

**Comment by smancill** [comment]
> If you’re implacably opposed I can attach another branch name, but it would require closing this PR and opening another so I’m not sure it’s exactly worth it.

No, it would be pointless. I don't have strong feelings in this case. We can merge this.

Now, if you ask me about that jj tool design choices (which I didn't know), then I would start arguing that having developed a tool that puts random branches names is not so useful and not very Git-compatible 🤭

**Review by smancill — APPROVED** [review]

**Comment by emilazy** [comment]
Back to a draft again as the `libresample` PR was merged upstream; will update this as soon as there’s a tag and get rid of those patches 😅

**Comment by emilazy** [comment]
Okay, I don’t know how long it’ll be before upstream cuts a release and this is one of the few things remaining for the big FFmpeg 4 PR so I’ve just bumped it to drop the vendored Meson patch for libresample but keeping the smaller tests one for now. Now it just needs merging before any further developments happen!

**Comment by smancill** [comment]
Result of `nixpkgs-review pr 332035` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>zrythm</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>kissfft</li>
    <li>kissfft.bin</li>
    <li>kissfft.dev</li>
    <li>kissfftFloat</li>
    <li>kissfftFloat.bin</li>
    <li>kissfftFloat.dev</li>
    <li>libresample</li>
    <li>libresample.bin</li>
    <li>libresample.dev</li>
    <li>musly</li>
    <li>musly.bin</li>
    <li>musly.dev</li>
    <li>musly.doc</li>
    <li>qm-dsp</li>
  </ul>
</details>

**Review by smancill — APPROVED** [review]

File: pkgs/by-name/li/libresample/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/li/libresample/package.nix]: Right now meta.licenses is always the collection of all licenses the project uses and we don't reflect the details. Anything else is done wrong.

Comment (emilazy) [reply pkgs/by-name/li/libresample/package.nix]: I’m not sure you understand how dual licensing works: any downstream can pick any licence, at their option. libresample can be used entirely under the terms of a BSD licence without any consideration of the LGPL terms. Therefore simply declaring BSD is entirely correct: Nixpkgs has the option to choose that licence, is exercising it, and neither Nixpkgs nor any package depending on this library is under any LGPL‐related obligations.

Consider a program that is available under either the GPL or a proprietary, non‐free licence. There are many historical examples of this, for instance Qt. Do you really think that it ought to be declared `licenses = [ lib.licenses.gpl3Only lib.licenses.unfreeRedistributable ]` and flag up unfree warnings, when the whole point is that anyone who receives the software can choose to use it under the terms of the GPL?

**Pull request merged**
